kube-core/src/request.rs

```diff
@@ -745,10 +739,58 @@ mod test {
     //}
 
     #[test]
-    fn watches_cannot_have_limits() {
-        let lp = ListParams::default().limit(5);
+    fn list_pods_from_cache() {
+        let url = corev1::Pod::url_path(&(), Some("ns"));
+        let gp = ListParams::default()
+            .resource_version("0")
+            .resource_version_match(ResourceVersionMatch::NotOlderThan);
```

Comment:
One point from reading the [spec](https://kubernetes.io/docs/reference/using-api/api-concepts/#semantics-for-get-and-list); and seeing the validation tests here; users are always required to set a resource version when using a match param. By having the two builder fns be separate we create the possibility for configuring only one of them (which, granted is caught by the validator).

One less error-prone way for users here is to not have the new `resource_version_match` and `resource_version` builders in the current state, and instead replace them with specific variants:

- `ListParams::not_older_than(mut self, ver: &str)`
- `ListParams::exact(mut self, ver: &str)`
- `ListParams::legacy(mut self, ver: &str)` (not setting match)

..which we can also encapsulate as a combined enum:

```rust
enum VersionMatch {
  Legacy
  Exact(String)
  #[default]
  NotOlderThan(String)
}
```

if we have this as a tri-state enum, then maybe it's enough to have a single new setter instead of the first proposed three:

- `ListParams::version_match(mut self, vmatch VersionMatch)`

At any rate, i think combining these two options can help us nudge users towards a default (and the spec says you should always set it, therefore the old is legacy and should not be our default).

Reply:
Great, it seems reasonable to me. I will arrange the code accordingly.

Reply:
P.S. one more idea. Maybe we should expand the enum to four options (and forbid setting resourceVersion="0" at all)?

```
enum VersionMatch {
  Any
  MostRecent
  #[default]
  Exact(String)
  NotOlderThan(String)
}
```

Reply:
I do like that bigger enum as it serves as a semantic interface to the options, and it leaves a much better prettier and non-breaking default (whereas we can change semantics in `watcher` that has a particular usage in mind). It does stray a bit more from "just model the api" that kube-client and kube-core is meant to do, but it is a mostly straight-forward remap to help avoid users have to deal with that huge table.

We would need to document the various enum options well (perhaps with the semantics text from the spec) clearly indicating what we are doing with resourceVersion etc, and also note in the `Exact`/`NotOlderThan` variants that there are special cases to be handled when paging (which also is something we also have to deal with in the watcher if we expose the same enum on `watcher::Config` - and we probably should).

> P.S.S. why is the resource version not a number?

sig-apimachinery does not want the integer to be consider as a public api and they are continually considering changing it. somewhere in the kubernetes website says it's an implementation detail and it must not be treated as anything other than a string.

Reply:
I refactored the code according to this dicsussion.

Reply:
Looking great!